src/torrent_info.cpp

```diff
@@ -980,12 +1107,47 @@ namespace {
 		swap(m_info_section, ti.m_info_section);
 		swap(m_piece_hashes, ti.m_piece_hashes);
 		m_info_dict.swap(ti.m_info_dict);
-		swap(m_merkle_tree, ti.m_merkle_tree);
 		swap(m_info_section_size, ti.m_info_section_size);
-		swap(m_merkle_first_leaf, ti.m_merkle_first_leaf);
 		swap(m_flags, ti.m_flags);
 	}
 
+	aux::vector<aux::vector<sha256_hash>, file_index_t>& torrent_info::merkle_trees()
+	{
+		TORRENT_ASSERT(m_merkle_trees.end_index() <= orig_files().end_file());
+		if (m_merkle_trees.empty())
+			m_merkle_trees.resize(orig_files().num_files());
+
+		for (file_index_t i{ 0 }; i < m_merkle_trees.end_index(); ++i)
+		{
+			if (orig_files().pad_file_at(i)) continue;
+			auto& f = m_merkle_trees[i];
+			if (f.empty() && orig_files().file_size(i) > 0)
+			{
+				auto const leafs = merkle_num_leafs(orig_files().file_num_blocks(i));
+				f.resize(merkle_num_nodes(leafs));
+				f[0] = orig_files().root(i);
+			}
+		}
+
+		return m_merkle_trees;
+	}
+
+	aux::vector<sha256_hash>& torrent_info::file_merkle_tree(file_index_t file) const
```

Comment:
I notice this function is `const`. this works because `m_merkle_trees` is `mutable`. Is this the only reason `m_merkle_trees` has to be mutable? the function above, is *not* `const` (even though I suspect it could be).

If it's possible, I would prefer to make this non-const, and make `m_merkle_trees` non-mutable. Since `torrent_info` *is* shared between threads, this is a race condition asking to happen.